src/lem/circuit.rs

```diff
@@ -552,26 +646,197 @@ impl Func {
+                            b_is_zero,
+                            &one,
+                            b_num,
+                        )?;
+
+                        let quotient = div(&mut cs.namespace(|| "quotient"), a_num, &divisor)?;
+
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, quotient);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Lt(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let (preallocated_preimg, lt) =
+                            &g.preallocated_less_than_slots[next_slot.consume_less_than()];
+                        for (i, n) in [a.hash(), b.hash()].into_iter().enumerate() {
+                            implies_equal(
+                                &mut cs.namespace(|| {
+                                    format!("implies equal for component {i} (OP {:?})", &op)
+                                }),
+                                not_dummy,
+                                n,
+                                &preallocated_preimg[i],
+                            )?;
+                        }
+                        let c = AllocatedPtr::from_parts(tag, lt.clone());
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Trunc(tgt, a, n) => {
+                        let b = u64::wrapping_sub(u64::wrapping_pow(2, *n), 1);
+                        let a = bound_allocations.get(a)?;
+                        let a_bits = a
+                            .hash()
+                            .to_bits_le_strict(&mut cs.namespace(|| "bitwise_and"))?;
+                        let mut trunc_bits = Vec::with_capacity(64);
+                        let mut b_rest = b;
+                        for a_bit in a_bits {
+                            let b_bit = b_rest & 1;
+                            if b_bit == 1 {
+                                trunc_bits.push(a_bit.clone());
+                            } else {
+                                trunc_bits.push(Boolean::Constant(false))
+                            }
+                            b_rest >>= 1;
+                        }
+                        let trunc = AllocatedNum::alloc(cs.namespace(|| "trunc"), || {
+                            let val = a
+                                .hash()
+                                .get_value()
+                                .map(|a| F::from_u64(a.to_u64_unchecked() & b))
+                                .unwrap();
```

Comment:
I think this `unwrap()` is going to cause a problem when you're trying to synthesize the shape and don't have any value for `a` — but still need to construct constraints that will work for any `a` that may be supplied when generating a witness.